### PR TITLE
Tagify

### PR DESCRIPTION
**Pull request by VaiTon**
Remove tagsInput (jQuery plugin) and use vanilla js [Tagify](https://github.com/yairEO/tagify).

Related to #2675

### CODE REVIEW

**Comment by stephanegigandet** [comment]
Hi @VaiTon , does it keep our autocomplete with the taxonomy entries + recently used entries?

**Comment by VaiTon** [comment]
> 
> 
> Hi @VaiTon , does it keep our autocomplete with the taxonomy entries + recently used entries?

For now it's just for taxonomy entries. Recently used ones are not implemented yet.

**Comment by hangy** [comment]
Wow, you're fast! 👍

**Comment by VaiTon** [comment]
> 
> 
> Hi @VaiTon , does it keep our autocomplete with the taxonomy entries + recently used entries?

Now implemented in ec6ac2f42. Not tested though as atm I can't seem to start docker.

**Comment by hangy** [comment]
> I can add tags (also from autocomplete) just fine, but get the following error if I try to edit a product that contains tags:

I've fixed that issue, but the stored recently used entries aren't used, as there is no suggestions displayed before someone starts typing. (And when they start typing the 'input' event handler is run.)

**Comment by hangy** [comment]
> I've fixed that issue, but the stored recently used entries aren't used, as there is no suggestions displayed before someone starts typing. (And when they start typing the 'input' event handler is run.)

I think this is fixed in f6f2cc0217ecad2607fff661703663efe65d45c1. I also took the liberty of moving the Tagify initialization to an external JavaScript file in 4b56a7b1bb67208bfad6236036fd8ddee986d968. Are you OK with these changes?

**Comment by stephanegigandet** [comment]
I deployed it on dev, (e.g. https://uk.openfoodfacts.dev ) but I have an error in the JS which is not handled:

image

That results in the product edit form scripts not being loaded / working.

It's this line:

const obj = JSON.parse(window.localStorage.getItem("po_last_tags"));

**Comment by stephanegigandet** [comment]
The suggestions are not working on the dev server, I think it's because of the authentication.
Can we change the host for the suggests request? Use the current host instead of world.
https://world.openfoodfacts.dev/cgi/suggest.pl?lc=en&tagtype=categories&term=c

**Comment by stephanegigandet** [comment]
I like the way the new tags look. :)

**Comment by hangy** [comment]
> I deployed it on dev, (e.g. https://uk.openfoodfacts.dev ) but I have an error in the JS which is not handled:
> 
> image
> 
> That results in the product edit form scripts not being loaded / working.
> 
> It's this line:
> 
> const obj = JSON.parse(window.localStorage.getItem("po_last_tags"));

Apparently, your browser's local storage might be corrupted: https://stackoverflow.com/q/18877643/11963 According to most sources I could find, `NS_ERROR_FILE_CORRUPTED` isn't related to the JavaScript source itself.

> The suggestions are not working on the dev server, I think it's because of the authentication.
> Can we change the host for the suggests request? Use the current host instead of world.
> https://world.openfoodfacts.dev/cgi/suggest.pl?lc=en&tagtype=categories&term=c

Sure. I guess that shouldn't change the functionality at all. I applied this in 1a7e05d085715bc9ac7f63e375239653b8746af9.

> I like the way the new tags look. :)

👍

**Comment by stephanegigandet** [comment]
> Apparently, your browser's local storage might be corrupted: https://stackoverflow.com/q/18877643/11963 According to most sources I could find, `NS_ERROR_FILE_CORRUPTED` isn't related to the JavaScript source itself.

Right, my point is that we should handle it gracefully. Right now it completely breaks the product edit form for me, I can't use it.

The stackoverflow post you link to suggests something like that:

try {
    setLocalStorageItem(key, value);
} catch(e) {
    if(e.name == "NS_ERROR_FILE_CORRUPTED") {
        showMessageSomehow("Sorry, it looks like your browser storage has been corrupted. Please clear your storage by going to Tools -> Clear Recent History -> Cookies and set time range to 'Everything'. This will remove the corrupted browser storage across all sites.");
    }
}

-> Quite frankly in this case, I would just disable that feature completely, instead of showing a message to the user. It's not that important a feature. But it clearly shouldn't break the whole product edit form.

**Comment by stephanegigandet** [comment]
@hangy @VaiTon : I made a fix to make the product edit form work on my apparently broken firefox: https://github.com/openfoodfacts/openfoodfacts-server/pull/2710/commits/ec4a7072aa0d245dd7f296c2a50478a97163bbf7
I'm clearly not experimented with javascript so please review. :)

File: cgi/product_multilingual.pl [hunk 1]
```diff
@@ -670,16 +670,17 @@ ($$)
...
-	let obj = JSON.parse(window.localStorage.getItem("po_last_tags"));
+${field}.on('add', function(e) {
+	const obj = JSON.parse(window.localStorage.getItem("po_last_tags"));
 
-	if (obj == null) {
+	const tag = e.detail.data.value;
+	if (obj === null) {
 		obj = {"${field}": [tag]}
-	} else if (obj["${field}"] == null) {
+	} else if (obj["${field}"] === null) {
```
Comment (VaiTon) [inline cgi/product_multilingual.pl]: Here i would also like to check if it's = to "" or to undefined.
>There are only six falsy values in JavaScript you should be aware of:
    false — boolean false
    0 — number zero
    “” — empty string
    null
    undefined
    NaN — Not A Number

File: html/js/product-multilingual.js [hunk 2]
```diff
@@ -494,8 +494,111 @@ function update_display(imagefield, first_display) {
...
+  let obj
+    try {
+      obj = JSON.parse(window.localStorage.getItem("po_last_tags"));
+    } catch(e) {
+      if(e.name == "NS_ERROR_FILE_CORRUPTED") {
+        obj = null;
+      }
+    }
+
+  if (obj === null) {
```
Comment (VaiTon) [inline html/js/product-multilingual.js]:
```suggestion
```

File: html/js/product-multilingual.js [hunk 3]
```diff
@@ -494,8 +494,111 @@ function update_display(imagefield, first_display) {
...
+    return [];
```
Comment (VaiTon) [inline html/js/product-multilingual.js]:
```suggestion
```

File: html/js/product-multilingual.js [hunk 4]
```diff
@@ -494,8 +494,111 @@ function update_display(imagefield, first_display) {
...
+  }
```
Comment (VaiTon) [inline html/js/product-multilingual.js]:
```suggestion
```

File: html/js/product-multilingual.js [hunk 5]
```diff
@@ -494,8 +494,111 @@ function update_display(imagefield, first_display) {
...
+
+  return obj[tagfield] === null ? [] : obj[tagfield];
```
Comment (VaiTon) [inline html/js/product-multilingual.js]:
```suggestion
  return obj[tagfield] == null ? [] : obj[tagfield];
```

File: html/js/product-multilingual.js [hunk 6]
```diff
@@ -494,8 +494,111 @@ function update_display(imagefield, first_display) {
...
+      }
+
+      obj[el.id].unshift(tag);
+    }
+
+
+    try {
+      window.localStorage.setItem("po_last_tags", JSON.stringify(obj));
+    } catch(e) {
+      if(e.name == "NS_ERROR_FILE_CORRUPTED") {
```
Comment (VaiTon) [inline html/js/product-multilingual.js]:
```suggestion
      if(err.name == "NS_ERROR_FILE_CORRUPTED") {
        // TODO: Tell the user
```

File: html/js/product-multilingual.js [hunk 7]
```diff
@@ -494,8 +494,111 @@ function update_display(imagefield, first_display) {
...
+        obj[el.id].pop();
+      }
+
+      obj[el.id].unshift(tag);
+    }
+
+
+    try {
+      window.localStorage.setItem("po_last_tags", JSON.stringify(obj));
+    } catch(e) {
```
Comment (VaiTon) [inline html/js/product-multilingual.js]:
```suggestion
    } catch(err) {
```

File: html/js/product-multilingual.js [hunk 8]
```diff
@@ -494,8 +494,111 @@ function update_display(imagefield, first_display) {
...
+    }
+  });
+
+  input.on('add', function(e) {
+    let obj;
+
+    try {
+      obj = JSON.parse(window.localStorage.getItem("po_last_tags"));
+    } catch(e) {
+      if(e.name == "NS_ERROR_FILE_CORRUPTED") {
```
Comment (VaiTon) [inline html/js/product-multilingual.js]:
```suggestion
      if(err.name == "NS_ERROR_FILE_CORRUPTED") {
```

File: html/js/product-multilingual.js [hunk 9]
```diff
@@ -494,8 +494,111 @@ function update_display(imagefield, first_display) {
...
+      });
+    }
+  });
+
+  input.on('add', function(e) {
+    let obj;
+
+    try {
+      obj = JSON.parse(window.localStorage.getItem("po_last_tags"));
+    } catch(e) {
```
Comment (VaiTon) [inline html/js/product-multilingual.js]:
```suggestion
    } catch(err) {
```

File: html/js/product-multilingual.js (hunk 5 above)
Comment (hangy) [reply html/js/product-multilingual.js]: The current eslint ruleset (based on some standard that I don't remember 🤷‍♂) would complain about the missing third `=`: https://eslint.org/docs/rules/no-eq-null

File: html/js/product-multilingual.js [hunk 10]
```diff
@@ -494,8 +494,111 @@ function update_display(imagefield, first_display) {
...
+    input.settings.whitelist = obj[el.id]; // reset the whitelist
+  });
+
+  document.getElementById('product_form').addEventListener('submit', function() {
+    el.value = input.value.map((obj) => obj.value).join(',');
+  });
+}
+
+function get_recents(tagfield) {
+  let obj
```
Comment (hangy) [inline html/js/product-multilingual.js]:
```suggestion
  let obj;
```

File: html/js/product-multilingual.js (hunk 5 above)
Comment (VaiTon) [reply html/js/product-multilingual.js]: If we want to keep that rule just create an if with === null or === undefined or === ""

File: html/js/product-multilingual.js (hunk 2 above)
Comment (hangy) [reply html/js/product-multilingual.js]: Wouldn't `return obj[tagfield] == null ? [] : obj[tagfield];` fail if `obj` was null?

File: html/js/product-multilingual.js [hunk 11]
```diff
@@ -494,8 +494,112 @@ function update_display(imagefield, first_display) {
...
+
+      obj[el.id].unshift(tag);
+    }
+
+
+    try {
+      window.localStorage.setItem("po_last_tags", JSON.stringify(obj));
+    } catch(err) {
+      if(err.name == "NS_ERROR_FILE_CORRUPTED") {
+        // TODO: Tell the user
```
Comment (hangy) [inline html/js/product-multilingual.js]: This shouldn't be left as a `TODO`.

**Comment by hangy** [comment]
> @hangy @VaiTon : I made a fix to make the product edit form work on my apparently broken firefox: [ec4a707](https://github.com/openfoodfacts/openfoodfacts-server/commit/ec4a7072aa0d245dd7f296c2a50478a97163bbf7)

Is it really useful to catch this and pollute the code with this error handling if it only affects a really small portion of users (where some Firefox profile got broken)?

**Comment by VaiTon** [comment]
> 
> 
> > @hangy @VaiTon : I made a fix to make the product edit form work on my apparently broken firefox: [ec4a707](https://github.com/openfoodfacts/openfoodfacts-server/commit/ec4a7072aa0d245dd7f296c2a50478a97163bbf7)
> 
> Is it really useful to catch this and pollute the code with this error handling if it only affects a really small portion of users (where some Firefox profile got broken)?

We could just surround the function with a catcher. In that way the code would be more readable and the user would still be alerted.

**Comment by stephanegigandet** [comment]
> Is it really useful to catch this and pollute the code with this error handling if it only affects a really small portion of users (where some Firefox profile got broken)?

Well we have no idea how many users are affected or could be affected in the future, and the consequence is very bad: it makes the product edit form not usable at all. So yes, I do believe we need a way to make sure the product edit form works.

About the alert to the user: I'm not sure if that's a good idea. This is a very small feature (saving the last tags used), I don't think it's a good idea to pop up an alert that could confuse users. Let's just turn the feature off if for some reason localstorage is not available.

File: html/js/product-multilingual.js (hunk 11 above)
Comment (stephanegigandet) [reply html/js/product-multilingual.js]: Let's just turn the feature off whenever there is an exception, without telling the user.

File: html/js/product-multilingual.js (hunk 2 above)
Comment (VaiTon) [reply html/js/product-multilingual.js]: You're right, I'll re-review this.

**Pull request merged**